dartagnan/src/main/java/com/dat3m/dartagnan/parsers/program/visitors/VisitorLitmusVulkan.java

```diff
@@ -0,0 +1,294 @@
+        return programBuilder.addChild(mainThread, EventFactory.newLocal(register, constant));
+    }
+
+    @Override
+    public Object visitLoadLocation(LitmusVulkanParser.LoadLocationContext ctx) {
+        Register register = (Register) ctx.register().accept(this);
+        MemoryObject location = programBuilder.getOrNewMemoryObject(ctx.location().getText());
+        Boolean atomatic = ctx.atomatic().isAtomic;
+        String mo = ctx.mo().content;
+        String scope = ctx.scope().content;
+        String storageClass = ctx.storageClass().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        String avvisSemantic = ctx.avvisSemantic().content;
+        Load load = EventFactory.newLoadWithMo(register, location, mo);
+        tagChecker(load, atomatic, mo, scope, avvisSemantic);
+        load.addTags(scope, storageClass, classSemantic);
+        return programBuilder.addChild(mainThread, load);
+    }
+
+    @Override
+    public Object visitRmwConstant(LitmusVulkanParser.RmwConstantContext ctx) {
+        Register register = (Register) ctx.register().accept(this);
+        MemoryObject location = programBuilder.getOrNewMemoryObject(ctx.location().getText());
+        IConst constant = (IConst) ctx.constant().accept(this);
+        Boolean atomatic = ctx.atomatic().isAtomic;
+        String mo = ctx.mo().content;
+        String scope = ctx.scope().content;
+        String storageClass = ctx.storageClass().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        String avvisSemantic = ctx.avvisSemantic().content;
+        VulkanRMW rmw = EventFactory.Vulkan.newRMW(location, register, constant, mo, scope);
+        rmw.addTags(scope, classSemantic, avvisSemantic);
+        tagChecker(rmw, atomatic, mo, scope, avvisSemantic);
+        rmw.addTags(scope, storageClass, classSemantic);
+        return programBuilder.addChild(mainThread, rmw);
+    }
+
+    @Override
+    public Object visitMemoryBarrier(LitmusVulkanParser.MemoryBarrierContext ctx) {
+        String mo = ctx.mo().content;
+        String scope = ctx.scope().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        Fence fence = EventFactory.newFence(ctx.getText().toLowerCase());
+        tagChecker(fence, false, mo, scope, "");
+        fence.addTags(scope, classSemantic);
+        return programBuilder.addChild(mainThread, fence);
+    }
+
+    @Override
+    public Object visitControlBarrier(LitmusVulkanParser.ControlBarrierContext ctx) {
+        String scope = ctx.scope().content;
+        String mo = ctx.mo().content;
+        String classSemantic = ctx.storageClassSemantic().content;
+        Expression fenceId = (Expression) ctx.barID().accept(this);
+        Fence fence = EventFactory.PTX.newFenceWithId(ctx.getText().toLowerCase(), fenceId);
+        fence.addTags(scope, mo, classSemantic);
+        return programBuilder.addChild(mainThread, fence);
```

Comment:
CBAR instructions are tagged as fences F [only if they are acq/rel](https://github.com/KhronosGroup/Vulkan-MemoryModel/blob/66c31d9cfcc03b1f5231779fbf3c42dc983b62e2/alloy/litmus.cpp#L187-L190). Every `Fence` event has the F tag, so you need to remove it if the instruction is not acq/rel

Reply:
updated

Reply:
what about `ACQ_REL`?

Reply:
Right, I have forgotten.. added now.